include/rocksdb/file_system.h

```diff
@@ -90,7 +90,17 @@ struct IOOptions {
   // Type of data being read/written
   IOType type;
 
-  IOOptions() : timeout(0), prio(IOPriority::kIOLow), type(IOType::kUnknown) {}
+  // Force directory fsync, some file systems like btrfs may skip directory
+  // fsync, set this to force the fsync
+  bool force_dir_fsync;
```

Comment:
I can sort of see ignoring `force_dir_fsync` to mean just do the fsync if I strain to think about it (because after all we're in a function called `Fsync()`...). So it's fine to me.

Reply:
I agree. Are you worried about the size of `IOOptions` to suggest having the `contents_synced` at argument? If it's not a concern, how about adding `contents_synced` to `IOOptions`, maybe also having the `string renamed_file` in `IOOptions` (to get ride of `FsyncAfterFileRename()`). Then we don't need to change public API. For existing POSIX filesystems, they just ignore these new added fields. And for btrfs, it can take advantage of that (or ignore that and just do dir-fsync is fine):
```
Fsync(opts) {
  // if it's btrfs, otherwise just do fsync dir:
  if (!opts.contents_synced) {
    if (opts.renamed_file) {
       Fsync(opts.renamed_file); // fsync renamed file
    } else {
       Fsync(fd_); // fsync dir
    }
  }
}
```
`contents_synced` is set to false by default, so the default behavior is not changed.